src/api/java/mekanism/api/chemical/BasicChemicalTank.java

```diff
@@ -266,9 +467,60 @@
-    private final Predicate<Chemical> validator;
+    /**
+     * Creates a tank with a given capacity, extract predicate, insert predicate, validation predicate, attribute validator, and content listener.
+     *
+     * @param capacity           Tank capacity.
+     * @param canExtract         Extract predicate.
+     * @param canInsert          Insert predicate.
+     * @param validator          Validation predicate.
+     * @param attributeValidator Chemical Attribute Validator, or {@code null} to fall back to {@link ChemicalAttributeValidator#DEFAULT}.
+     * @param listener           Contents change listener.
+     *
+     * @since 10.7.11
+     */
+    public static IChemicalTank createModern(long capacity, BiPredicate<Holder<Chemical>, @NotNull AutomationType> canExtract,
+          BiPredicate<Holder<Chemical>, @NotNull AutomationType> canInsert, Predicate<Holder<Chemical>> validator,
+          @Nullable ChemicalAttributeValidator attributeValidator, @Nullable IContentsListener listener) {
+        if (capacity < 0) {
+            throw new IllegalArgumentException("Capacity must be at least zero");
+        }
+        Objects.requireNonNull(canExtract, "Extraction validity check cannot be null");
+        Objects.requireNonNull(canInsert, "Insertion validity check cannot be null");
+        Objects.requireNonNull(validator, "Chemical validity check cannot be null");
+        return new BasicChemicalTank(capacity, canExtract, canInsert, validator, attributeValidator, listener, null);
+    }
+
+    @SuppressWarnings("removal")
+    @Deprecated(forRemoval = true, since = "10.7.11")
+    private static BiPredicate<Chemical, @NotNull AutomationType> wrapAutomationPredicate(BiPredicate<Holder<Chemical>, @NotNull AutomationType> predicate) {
+        if (predicate == holderAlwaysTrueBi || predicate == holderInternalOnly || predicate == holderNotExternal || predicate == holderManualOnly) {
+            return (BiPredicate<Chemical, @NotNull AutomationType>) (BiPredicate<?, @NotNull AutomationType>) predicate;
+        }
+        return (chemical, automationType) -> predicate.test(chemical.getAsHolder(), automationType);
+    }
+
+    @Deprecated(forRemoval = true, since = "10.7.11")
+    private static BiPredicate<Holder<Chemical>, @NotNull AutomationType> wrapAutomationPredicateToModern(BiPredicate<Chemical, @NotNull AutomationType> predicate) {
+        if (predicate == alwaysTrueBi || predicate == internalOnly || predicate == notExternal || predicate == manualOnly) {
+            return (BiPredicate<Holder<Chemical>, @NotNull AutomationType>) (BiPredicate<?, @NotNull AutomationType>) predicate;
+        }
+        return (chemical, automationType) -> predicate.test(chemical.value(), automationType);
+    }
+
+    private final Predicate<Holder<Chemical>> validator;
+    //TODO - 1.22: Rename these to canExtract and canInsert
+    protected final BiPredicate<Holder<Chemical>, @NotNull AutomationType> canExtractModern;
```

Comment:
shouldn't we make these just use the stack?

Reply:
I was wondering about that, so I will go ahead and so so, even though we don't care about the actual amount and we don't have data components for chemicals.